.circleci/config.yml

```diff
@@ -1,106 +1,183 @@
-      - run:
-          name: Install python3
-          command: sudo apt install python3 python3-pip
-      - run:
-          name: Install dependencies
-          command: |
-            # Get rid of pyenv stuff
-            sudo rm -rf .pyenv/ /opt/circleci/.pyenv/
-            # Download and install miniconda
-            curl https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh  > miniconda.sh
-            bash miniconda.sh -b -p $HOME/miniconda
-            # Create an anaconda virtualenv for python ${CONDA_ENV_PY_VERSION} and make that the default python interpreter
-            echo ". /home/circleci/miniconda/etc/profile.d/conda.sh" >> $BASH_ENV
-            echo "conda deactivate; conda activate kedro-viz" >> $BASH_ENV
-            . /home/circleci/miniconda/etc/profile.d/conda.sh
-            conda create --name kedro-viz python=${CONDA_ENV_PY_VERSION} -y
-            source $BASH_ENV
-      - run:
-          name: Pip install dependencies
-          command: |
-            python -m pip install -U pip setuptools wheel
-            pip install git+https://github.com/quantumblacklabs/kedro
-            cd package && pip install -r requirements.txt -U
-            pip install -r test_requirements.txt -U
-            cd ..
-      - run:
-          name: Echo package versions
-          command: |
-              python -V
-              pip -V
-              kedro -V
-      - run:
-          name: Install Node dependencies
-          command: npm ci
-      - run:
-          name: Run Eslint
-          command: npm run lint
-      - run:
-          name: Run JavaScript tests
-          command: npm run build:css && npm test
-      - run:
-          name: Test lib transpilation
-          command: npm run lib
-      - run:
-          name: Run pylint and flake8
-          command: |
-            make pylint
-      - run:
-          name: Run Python tests
-          command: make pytest
-      - run:
-          name: Run all end to end tests
-          command: |
-            make e2e-tests
-      - run:
-          name: Cleanup
-          command: make clean
-      - save_cache:
-          paths:
-            - node_modules
-          key: v1-dependencies-{{ checksum "package.json" }}
-      - run:
-          name: Build application
-          command: npm run build
-      - run:
-          name: Package app
-          command: make package
-      - run:
-          name: Check for legal compliance
-          command: |
-            make legal
+      - restore_cache: *restore_cache
+      - run: *install_python3
+      - run: *install_python_environment
+      - run: *install_pip_dependencies
+      - run: *echo_package_versions
+      - run: *install_node_dependencies
+      - run: *run_eslint
+      - run: *run_javascript_tests
+      - run: *test_lib_transpilation
+      - run: *run_pylint_and_flake8
+      - run: *run_python_tests
+      - run: *run_e2e_tests
+      - run: *cleanup
+      - save_cache: *save_cache
```

Comment:
Wait, still you use the same set of steps for all builds, right? Which means you do `save_cache` three times. Doesn't it complain about it? I thought the entire idea of this refactoring was to make steps different for different versions.

Reply:
As we discussed offline: it's not persisting to workspace anymore, it's cache saving, which allows this multi-writing, and we don't care which one we actually receive in the next job.